Framework/HistogramData/inc/MantidHistogramData/VectorOf.h

```diff
@@ -0,0 +1,158 @@
+#ifndef MANTID_HISTOGRAMDATA_VECTOROF_H_
+#define MANTID_HISTOGRAMDATA_VECTOROF_H_
+
+#include "MantidHistogramData/DllConfig.h"
+#include "MantidKernel/cow_ptr.h"
+#include "MantidKernel/make_cow.h"
+
+namespace Mantid {
+namespace HistogramData {
+namespace detail {
+
+/** VectorOf
+
+  This class is an implementation detail of classes like HistogramData::BinEdges
+  and HistogramData::Points. It wraps a copy-on-write pointer to an underlying
+  data type based on std::vector, such as HistogramX and HistogramY.
+
+  The first template parameter is the type of the inheriting class. This is the
+  CRTP (curiously recurring template pattern).
+
+  The second template parameter is the type of the object to store.
+
+  @author Simon Heybrock
+  @date 2016
+
+  Copyright &copy; 2016 ISIS Rutherford Appleton Laboratory, NScD Oak Ridge
+  National Laboratory & European Spallation Source
+
+  This file is part of Mantid.
+
+  Mantid is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 3 of the License, or
+  (at your option) any later version.
+
+  Mantid is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+  File change history is stored at: <https://github.com/mantidproject/mantid>
+  Code Documentation is available at: <http://doxygen.mantidproject.org>
+*/
+template <class T, class CowType> class VectorOf {
```

Comment:
It is part of the CRTP. It ensures that subtypes like `Points` and `BinEdges` are different and incompatible types, by having separate base classes `VectorOf<Points>` and `VectorOf<BinEdges>.